README.md

```diff
@@ -6,7 +6,498 @@
+and then somehow feed the remaining bytes into the next stage of processing
+(e.g this might be a nested format or a different parsing branch in the formal
+ grammar of the protocol). Both of these scenarios require logic that is
+difficult to write correctly and results in unnecessary copying of the input
+bytes.
+
+A major difference in the FastStreams design is that the arbitrary-length
+data obtained from the input device is managed by the stream itself and you
+are provided with an API allowing you to control precisely how much data
+is consumed from the stream. Consuming the buffered content does not invoke
+costly asynchronous calls and you are allowed to peek at the stream contents
+before deciding which step to take next (something crucial for handling formal
+grammars). Thus, using the FastStreams API results in code that is both highly
+efficient and easy to author.
+
+### Higher efficiency is possible if we say goodbye to the good old single buffer.
+
+The buffering logic inside the stream divides the data into "pages" which
+are allocated with a known fast path in the Nim allocator and which can be
+efficiently transferred between streams and threads in the layered streams
+scenario or in IPC mechanisms such as `AsyncChannel`. The consuming code can
+be aware of this, but doesn't need to. The most idiomatic usage of the API
+handles the buffer switching logic automatically for the user.
+
+Nevertheless, the buffering logic can be configured for unbuffered reads
+and writes and it supports efficiently various common real-world patterns
+such as:
+
+* Length prefixes
+
+  To handle protocols with length prefixes without any memory overhead,
+  the output streams support "delayed writes" where a portion of the
+  stream content is specified only after the prefixed content is written
+  to the stream.
+
+* Block compressors and Block ciphers
+
+  These can benefit significantly from a more precise control over
+  the stride of the buffered pages which can be configured to match
+  the block size of the encoder.
+
+* Content with known length
+
+  Some streams have a known length which allows us to accurately estimate
+  the size of the transformed content. The `len` and `ensureRunway` APIs
+  make sure such cases are handled as optimally as possible.
+
+## Basic API usage
+
+The FastStreams API consists of ony few major object types:
+
+### `InputStream`
+
+An `InputStream` manages a particular input device. The library offers out
+of the box the following input stream types:
+
+* `fileInput`
+
+  For reading files through the familiar `fread` API from the C run-time.
+
+* `memFileInput`
+
+  For reading memory mapped files which provides best performance.
+
+* `unsafeMemoryInput`
+
+  For handling strings, sequences and openarrays as an input stream. <br />
+  You are responsible for ensuring that the backing buffer won't be invalidated
+  while the stream is being used.
+
+* `memoryInput`
```

Comment:
put this one above `unsafeMemoryInput`? it's simpler to use and a good default in nim's GC world

Reply:
But it's so must more inefficient! oh, well :)